jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java

```diff
@@ -571,61 +574,67 @@ static Chunk next(Chunk chunk)
 
         /**
          * <p>Returns a new {@code Chunk} whose {@code ByteBuffer} is a slice, with the given
          * position and limit, of the {@code ByteBuffer} of the source {@code Chunk} unless the
-         * source is {@link #isTerminal() terminal} in which case {@code this} is returned, or
-         * if {@code position == limit} in which case {@link #EOF} or {@link #EMPTY} is
+         * source {@link #hasRemaining() has no remaining bytes} in which case:</p>
+         * <ul>
+         * <li>{@code this} is returned if {@link #isLast()} is {@code true}</li>
+         * <li>{@link #EMPTY} is returned if {@link #isLast()} is {@code false}</li>
+         * </ul>
+         * <p>If {@code position == limit} then either {@link #EOF} or {@link #EMPTY} is
          * returned depending on the value of {@code last}.</p>
-         * <p>The returned {@code Chunk} retains the source {@code Chunk} and it is linked
-         * to it via {@link #from(ByteBuffer, boolean, Retainable)}.</p>
+         * <p>If the source {@code Chunk} has remaining bytes, the returned {@code Chunk}
+         * is linked to the source {@code Chunk} via
+         * {@link #from(ByteBuffer, boolean, Retainable)}.</p>
          *
          * @param position the position at which the slice begins
          * @param limit the limit at which the slice ends
          * @param last whether the new Chunk is last
-         * @return a new {@code Chunk} retained from the source {@code Chunk} with a slice
-         * of the source {@code Chunk}'s {@code ByteBuffer}
+         * @return a new {@code Chunk} with a slice of the source {@code Chunk}'s
+         * {@code ByteBuffer}
          */
         default Chunk slice(int position, int limit, boolean last)
         {
-            if (isTerminal())
-                return this;
-            if (position == limit)
-                return last ? EOF : EMPTY;
-            ByteBuffer sourceBuffer = getByteBuffer();
-            int sourceLimit = sourceBuffer.limit();
-            sourceBuffer.limit(limit);
-            int sourcePosition = sourceBuffer.position();
-            sourceBuffer.position(position);
-            ByteBuffer slice = sourceBuffer.slice();
-            sourceBuffer.limit(sourceLimit);
-            sourceBuffer.position(sourcePosition);
-            retain();
-            return from(slice, last, this);
+            if (hasRemaining())
+            {
+                if (position == limit)
+                    return last ? EOF : EMPTY;
+                ByteBuffer sourceBuffer = getByteBuffer();
+                int sourceLimit = sourceBuffer.limit();
+                sourceBuffer.limit(limit);
+                int sourcePosition = sourceBuffer.position();
+                sourceBuffer.position(position);
+                ByteBuffer slice = sourceBuffer.slice();
+                sourceBuffer.limit(sourceLimit);
+                sourceBuffer.position(sourcePosition);
+                return from(slice, last, this);
```

Comment:
If I'm  reading this correctly, the source buffer is momentarily having its limit and position changed, then a slice taken, then the pointers reverted.   This is not thread safe as another thread looking at the source buffer will momentarily see the wrong pointers.

Can't we slice (or duplicate) and then adjust the pointers?

Reply:
No we cannot because otherwise we are allowing the code that receives the slice to see bytes past the slice's position and limit.
We should never have 2 threads looking at the same buffer, and I believe the code is already built in a way to guarantee this.

Reply:
That is a very fragile assumption for a class in util. 

Perhaps we need to duplicate, adjust then slice?

Reply:
This is in jetty-io, not jetty-util.  does that change your opinion?

Reply:
Io is a "util" package. Big call to say that it's classes will never be used by multiple threads

Reply:
> Big call to say that it's classes will never be used by multiple threads

It's not. 
`ByteBuffer`s should not be used by multiple threads and `Chunk` is just a wrapper around a `ByteBuffer`, so inherits the same limitation.
Slices should not allow applications to look around for bytes that possibly don't belong to them, so they must be sliced at the right capacity or we have a security issue. We have been very careful elsewhere in doing that, and here is no different.

Reply:
We definitely have examples of ByteBuffers being used by multiple threads.  We have them in the HttpContent cache and multiple threads each slice off a buffer to use.
I don't see why we could not retain a Content.Chunk that we want to send to multiple destinations in multiple threads, each slicing off the chunk they want to send.

Reply:
So they are not used by multiple threads. 
They are just sliced by multiple threads, and then each thread uses its own slice.

The reason to not retain inside `slice()` is that `slice()` does not know what use you are doing with the slice, so it cannot just retain() a priori, only to be released 2 lines later -- you complained about this pattern before.

So `slice()` does not retain, and if the caller code wants to send it to multiple consumers, it will retain the slice as many times as it wants. How can `slice()` know how many times you want to retain it because you are sending it to more than one consumer?

Reply:
The point here is not about retain or release. The point is that slice is not thread safe. If each thread calls slice then they will get random results as they may see the pointers passed in by a different thread.

Reply:
> The point is that slice is not thread safe.

It's not supposed to be, never was, and never will.
I don't understand why you are making this argument of thread safety?

Reply:
I think this is a bug. But it's a preexisting big unrelated to this PR. I'll open another issue.

Reply:
I don't think it is a bug, because `Chunk`s are supposed to be used by 1 thread only.
I don't think we have a use case for Chunks used concurrently, so why do you think it is plausible?

We can make the same argument for `ByteBuffer` or `ArrayList` -- it's plausible to use them concurrently, but we don't make them thread safe because they are not supposed to be used concurrently.

Before the `Chunk` abstraction, in Jetty 11, we are using `ByteBuffer`s (in the client, directly; in the server, wrapped into a `HttpInput.Content`). None of these usages is concurrent, and the design and usage of the `ByteBuffer` never takes thread safety into account (so much so that `ByteBuffer` was used directly), so I'm not sure why now we should?

Reply:
We store ByteBuffers in the static content cache and multiple threads can concurrently  slice the same buffer to serve it. The buffer itself is not thread safe, but the slice method must be.

If multiple threads are not being used, then why slice? Just use the original chunk.

It's entirely plausible for a chunk to be retained and used in a similar way, so the slice method needs to be thread safe.

Reply:
I understand that you are proposing two choices:

1. keep a parameter-less `slice()` method similar to `ByteBuffer` and change the name of the `slice(int, int, boolean)` to `view(int, int, boolean)`, so that it's clearer that it is not thread safe; or
2. remove `slice(int, int, boolean)`, do the position/limit modification on the original chunk, and then call the parameter-less `slice()`

Did I get that correctly?

> If multiple threads are not being used, then why slice? Just use the original chunk.

We cannot because we will expose to applications the whole buffer (possibly with bytes belonging to the next request, or the next frames).

In Jetty 11 we expose the whole buffer up to `HttpInput`, but then applications only have one way of reading from it via `read(byte[])`, so the application could never read before or after the position/limit of the `ByteBuffer`.

In Jetty 12 we have a different API where we may expose the whole buffer directly to applications, so we slice to avoid that, even in the single thread case.

Reply:
I've not reviewed in enough detail to say that i have an actual prism that i like. At this stage so i can tell you it's what I don't like.

I don't like a method called slice that produces a dependent chunk that calling release on the original will release the slice, out that releasing the slice will also release the original.

The method you want us not slicing to create a new chunk that can be independently used. From the sound of it, it is temporarily restricting the range of the existing chunk, without creating a new reference, so a synchronous call can be made with that chunk to application coffee that we don't want it peeking at data outside the range.. But that synchronous call somehow knows not to release the chunk it is passed, but also will call retain itself if it is actually asynchronous.

I need to get on a proper computer and look in detail how you are using these methods to really say more (early next week).

I'm short, slice implies creating something independent, which you apparently don't need. You just need something that restricts the view.... but that can be retained if need be.

Reply:
I've opened #9141 as this bug is unrelated to this PR.